third_party/java/java_tools/update_java_tools.sh

```diff
@@ -0,0 +1,106 @@
+#
+# For usage please run
+# ~/third_party/java/java_tools/update_java_tools.sh help
+
+declare -a all_tools=("JavaBuilder" "VanillaJavaBuilder" "GenClass" "Runner" \
+"ExperimentalRunner" "JacocoCoverage" "Turbine" "TurbineDirect" "SingleJar")
+
+declare -A tool_name_to_target=( ["JavaBuilder"]="src/java_tools/buildjar:JavaBuilder_deploy.jar" \
+["VanillaJavaBuilder"]="src/java_tools/buildjar:VanillaJavaBuilder_deploy.jar" \
+["GenClass"]="src/java_tools/buildjar/java/com/google/devtools/build/buildjar/genclass:GenClass_deploy.jar" \
+["Runner"]="src/java_tools/junitrunner/java/com/google/testing/junit/runner:Runner_deploy.jar" \
+["ExperimentalRunner"]="src/java_tools/junitrunner/java/com/google/testing/junit/runner:ExperimentalRunner_deploy.jar" \
+["JacocoCoverage"]="src/java_tools/junitrunner/java/com/google/testing/coverage:JacocoCoverage_jarjar_deploy.jar" \
+["Turbine"]="src/java_tools/buildjar/java/com/google/devtools/build/java/turbine/javac:turbine_deploy.jar" \
+["TurbineDirect"]="src/java_tools/buildjar/java/com/google/devtools/build/java/turbine:turbine_direct_binary_deploy.jar" \
+["SingleJar"]="src/java_tools/singlejar/java/com/google/devtools/build/singlejar:bazel-singlejar_deploy.jar")
+
+usage="This script updates the checked-in jars corresponding to the tools "\
+"used by the Java rules in Bazel.
+
+To update all the tools simultaneously run from your bazel workspace root:
+~/third_party/java/java_tools/update_java_tools.sh
+
+To update only one or one subset of the tools run
+~/third_party/java/java_tools/update_java_tools.sh tool_1 tool_2 ... tool_n
+
+where tool_i can have one of the values ${all_tools[@]}
+
+For example, to update only JavaBuilder run
+~/third_party/java/java_tools/update_java_tools.sh JavaBuilder
+
+To update JavaBuilder, Turbine and SingleJar run
+~/third_party/java/java_tools/update_java_tools.sh JavaBuilder Turbine SingleJar
+"
+
+if [[ ! -z "$1" && $1 = "help" ]]; then
+  echo "$usage"
+  exit
+fi
+
+tools_to_update=()
+
+if [[ ! -z "$@" ]]
+then
+   # Update only the tools specified on the command line.
+   tools_to_update=("$@")
+else
+  # If no tools were specified update all of them.
+  tools_to_update=("${all_tools[@]}")
+fi
+
+
+updated_tools=()
+not_updated_tools=()
+
+function update_tool() {
+  local bazel_target="${1}"; shift
+  bazel build "$bazel_target"
+
+  local binary=$(echo "bazel-bin/$bazel_target" | sed 's@:@/@')
+
+  if [[ ! -f "$binary" ]]; then
+    binary=$(echo "bazel-genfiles/$bazel_target" | sed 's@:@/@')
+  fi
+
+  local tool_basename=$(basename $binary)
+  if [[ -f "$binary" ]]; then
+    cp -f "$binary" "third_party/java/java_tools/$tool_basename"
+    echo "Updated third_party/java/java_tools/$tool_basename"
+    updated_tools+=("third_party/java/java_tools/$tool_basename")
+  else
+    echo "Could not build $bazel_target"
+    not_updated_tools+=("third_party/java/java_tools/$tool_basename")
+  fi
+
+
+}
+
+for tool in "${tools_to_update[@]}"
```

Comment:
Opt: bash associative arrays are scary. How about simply

```
function process() {
  bazel build $2
  cp ... ...
}

process SingleJar src/java_tools/singlejar/java/com/google/devtools/build/singlejar:bazel-singlejar_deploy.jar
process GenClass rc/java_tools/buildjar/java/com/google/devtools/build/buildjar/genclass:GenClass_deploy.jar
```

More repetitive, but also more readable for mere mortals.

Reply:
Ah, this is exactly the way I started the script! However, it ended up having too much repetition especially because I wanted to be able to print the usage. Doing it this way requires repetition 3 times, and every time one of the targets or tool name changes we have to update the other 2  as well. I updated the documentation, does it make it more readable? I really want to avoid copying all these strings everywhere.

Reply:
I'd be OK with slightly worse usage text if it makes the script simpler.